File: app/src/main/java/com/asfoundation/wallet/repository/ErrorMapper.java
package com.asfoundation.wallet.repository;

import com.asfoundation.wallet.util.UnknownTokenException;
import java.net.UnknownHostException;

/**
 * Created by trinkes on 20/03/2018.
 */

public class ErrorMapper {

  public static final String INSUFFICIENT_ERROR_MESSAGE =
      "insufficient funds for gas * price + value";
  public static final String NONCE_TOO_LOW_ERROR_MESSAGE = "nonce too low";

  public PaymentTransaction.PaymentState map(Throwable throwable) {
    throwable.printStackTrace();
    if (throwable instanceof UnknownHostException) {
      return PaymentTransaction.PaymentState.NO_INTERNET;
    }
    if (throwable instanceof WrongNetworkException) {
      return PaymentTransaction.PaymentState.WRONG_NETWORK;
    }
    if (throwable instanceof TransactionNotFoundException) {
      return PaymentTransaction.PaymentState.ERROR;
    }
    if (throwable instanceof UnknownTokenException) {
      return PaymentTransaction.PaymentState.UNKNOWN_TOKEN;
    }
    if (throwable instanceof TransactionException) {
      switch (throwable.getMessage()) {
        case INSUFFICIENT_ERROR_MESSAGE:
          return PaymentTransaction.PaymentState.NO_FUNDS;
        case NONCE_TOO_LOW_ERROR_MESSAGE:
          return PaymentTransaction.PaymentState.NONCE_ERROR;
      }
    }
    return PaymentTransaction.PaymentState.ERROR;
  }
}

File: app/src/main/java/com/asfoundation/wallet/repository/ErrorMapper.kt
package com.asfoundation.wallet.repository

import com.appcoins.wallet.appcoins.rewards.getMessage
import com.asfoundation.wallet.repository.PaymentTransaction.PaymentState
import com.asfoundation.wallet.util.UnknownTokenException
import retrofit2.HttpException
import java.net.UnknownHostException

class ErrorMapper {
  fun map(throwable: Throwable): PaymentError {
    throwable.printStackTrace()
    return when (throwable) {
      is HttpException -> mapHttpException(throwable)
      is UnknownHostException -> PaymentError(PaymentState.NO_INTERNET)
      is WrongNetworkException -> PaymentError(PaymentState.WRONG_NETWORK)
      is TransactionNotFoundException -> PaymentError(PaymentState.ERROR)
      is UnknownTokenException -> PaymentError(PaymentState.UNKNOWN_TOKEN)
      is TransactionException -> mapTransactionException(throwable)
      else -> PaymentError(PaymentState.ERROR, null, throwable.message)
    }
  }

  private fun mapTransactionException(throwable: Throwable): PaymentError {
    return when (throwable.message) {
      INSUFFICIENT_ERROR_MESSAGE -> PaymentError(PaymentState.NO_FUNDS)
      NONCE_TOO_LOW_ERROR_MESSAGE -> PaymentError(PaymentState.NONCE_ERROR)
      else -> PaymentError(PaymentState.ERROR, null, throwable.message)
    }
  }

  private fun mapHttpException(exception: HttpException): PaymentError {
    return if (exception.code() == FORBIDDEN_CODE) {
      PaymentError(PaymentState.FORBIDDEN)
    } else {
      val message = exception.getMessage()
      PaymentError(PaymentState.ERROR, exception.code(), message)
    }
  }

  companion object {
    private const val INSUFFICIENT_ERROR_MESSAGE = "insufficient funds for gas * price + value"
    private const val NONCE_TOO_LOW_ERROR_MESSAGE = "nonce too low"
    private const val FORBIDDEN_CODE = 403
  }
}